src/daft-logical-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -416,6 +416,48 @@ impl PushDownProjection {
                     .or(Transformed::yes(new_plan));
                 Ok(new_plan)
             }
+            LogicalPlan::Union(union) => {
+                if !union.is_all {
+                    // can not push down past a DISTINCT
+                    return Ok(Transformed::no(plan));
+                }
+
+                // Get required columns from projection and upstream.
+                let combined_dependencies = plan
+                    .required_columns()
+                    .iter()
+                    .flatten()
+                    .chain(upstream_plan.required_columns().iter().flatten())
+                    .cloned()
+                    .collect::<IndexSet<_>>();
+
+                // Skip optimization if no columns would be pruned.
+                let grand_upstream_plan = &upstream_plan.children()[0];
+                let grand_upstream_columns = grand_upstream_plan.schema().names();
+                if grand_upstream_columns.len() == combined_dependencies.len() {
+                    return Ok(Transformed::no(plan));
+                }
+
+                let pushdown_column_exprs: Vec<ExprRef> = combined_dependencies
+                    .into_iter()
+                    .map(col)
+                    .collect::<Vec<_>>();
+                let new_left_subprojection: LogicalPlan =
+                    { Project::try_new(union.lhs.clone(), pushdown_column_exprs.clone())?.into() };
+                let new_right_subprojection: LogicalPlan =
+                    { Project::try_new(union.rhs.clone(), pushdown_column_exprs)?.into() };
+
+                let new_upstream = upstream_plan.with_new_children(&[
+                    new_left_subprojection.into(),
+                    new_right_subprojection.into(),
+                ]);
+                let new_plan = Arc::new(plan.with_new_children(&[new_upstream.into()]));
+                // Retry optimization now that the upstream node is different.
+                let new_plan = self
+                    .try_optimize_node(new_plan.clone())?
+                    .or(Transformed::yes(new_plan));
+                Ok(new_plan)
+            }
```

Comment:
Will we actually ever hit this, since union is always turned into different ops?

Reply:
i was thinking that too. I don't think it'll ever hit right now. Will remove.